src/content/en/updates/2018/06/audio-worklet-design-pattern.md

```diff
@@ -0,0 +1,414 @@
+SharedArrayBuffer and the main thread. The following steps describe what should
+happen in the initialization phase.
+
+##### Initialization
+
+  1. [Main] AudioWorkletNode constructor gets called.
+    1. Create Worker.
+    1. The associated AudioWorkletProcessor will be created.
+  1. [DWGS] Worker creates 2 SharedArrayBuffers. (one for shared states and the
+    other for audio data)
+  1. [DWGS] Worker sends SharedArrayBuffer references to AudioWorkletNode.
+  1. [Main] AudioWorkletNode sends SharedArrayBuffer references to
+    AudioWorkletProcessor.
+  1. [AWGS] AudioWorkletProcessor notifies AudioWorkletNode that the setup is
+    completed.
+
+Once the initialization is completed, `AudioWorkletProcessor.process()` starts
+to get called. The following is what should happen in each iteration of the
+rendering loop.
+
+##### Rendering Loop
+
+<figure>
+  <img
+      class="diagram"
+      src="/web/updates/images/2018/06/aw-design-pattern-7.png">
+  <figcaption>
+    Multi-threaded rendering with SharedArrayBuffers
+  </figcaption>
+</figure>
+
+  1. [AWGS] `AudioWorkletProcessor.process(inputs, outputs)` gets called for
+    every render quantum.
+    1. `inputs` will be pushed into **Input SAB**.
+    1. `outputs` will be filled by consuming audio data in **Output SAB**.
+    1. Updates **States SAB** with new buffer indexes accordingly.
+    1. If **Output SAB** gets close to underflow threshold, Wake Worker to
+      render more audio data.
+  1. [DWGS] Worker waits (sleeps) for the wake signal from
+    `AudioWorkletProcessor.process()`. When it wakes up:
+    1. Fetches buffer indexes from **States SAB**.
+    1. Run the process function with data from **Input SAB** to fill
+      **Output SAB**.
+    1. Updates **States SAB** with buffer indexes accordingly.
+    1. Goes to sleep and wait for the next signal.
+
+The example code can be found here, but note that the SharedArrayBuffer
+experimental flag must be enabled for this demo to work. The code was written
+with pure JS code for simplicity, but it can be replaced with WebAssembly code
+if needed. Such case should be handled with extra care by wrapping memory
+management with
+[HeapAudioBuffer](https://github.com/GoogleChromeLabs/web-audio-samples/blob/gh-pages/audio-worklet/design-pattern/lib/wa2.js#L34)
+class.
+
+## Conclusion
+
+The ultimate goal of the AudioWorklet is to make the Web Audio API truly
+"extensible". The multi-year effort went to its design to make it possible to
+implement the rest of Web Audio API with the AudioWorklet. In turn, now we have
+the higher complexity in its design and it can be a unexpected challenge.
+
+Fortunately, the reason for such complexity is purely to empower developers.
+Being able to run WebAssembly on AudioWorkletGlobalScope unlocks huge potential
+for high-performance audio processing on the web. For large-scale audio
+applications written in C or C++, using the AudioWorklet with SharedArrayBuffer
```

Comment:
"using the...SharedArrayBuffer" -> "using an AudioWorklet with SharedArrayBuffers"?